File: packages/webcomponents/src/switcher/avatar-expression.ts
export type AgentActivity = 'idle' | 'thinking' | 'working' | 'awaiting';

export const ACTIVITIES: readonly AgentActivity[] = ['idle', 'thinking', 'working', 'awaiting'];

export interface GazePoint {
  readonly x: number;
  readonly y: number;
}

export interface BrowPose {
  readonly raise: number;

  readonly tilt: number;
}

export interface BrowPair {
  readonly left: BrowPose;
  readonly right: BrowPose;
}

export const EYE_R = 38;
export const EYE_CY = 50;
export const LEFT_CX = 55;
export const RIGHT_CX = 145;
export const PUPIL_R = 18;
export const MAX_OFFSET = 16;

export const SOCKET_MIN_RX = 10;

export const PUPIL_MIN_FRACTION = 0.22;

export const SHAPE_EASE = 6;

export const BLINK_IN_MS = 110;
export const BLINK_OUT_MS = 130;

export const BLINK_APEX_MS = 120;
export const BLINK_SQUISH = 0.08;
export const BLINK_PERIOD_LEFT_MS = 3400;
export const BLINK_PERIOD_RIGHT_MS = 4600;

export const LID_EASE = 5;

export const GLOWER_LID = 0.38;
export const GLOWER_MS = 2600;

export const SCRUTINY_LID = 0.22;
export const SCRUTINY_MS = 1000;
export const DROWSE_START_LID = 0.1;
export const DROWSE_END_LID = 0.55;
export const DROWSE_RAMP_S = 12;
export const DEFAULT_DROWSE_DELAY_S = 90;

export const LID_OPEN_EPSILON = 0.001;

export const LID_LINE_EPSILON = 0.02;

export const LID_OVERSHOOT = 3;

export const POP_MS = 350;
export const POP_GAIN = 0.16;

export const BROW_HALF_WIDTH = 22;
export const BROW_Y = 2;
export const BROW_STROKE = 8;
export const BROW_TRANSITION_MS = 350;

export const RECOCK_FLIP_CHANCE = 0.65;

export const BASE_BROWS: BrowPair = {
  left: { raise: -9, tilt: -10 },
  right: { raise: 2, tilt: 6 },
};

export const SACCADE_TARGETS: readonly GazePoint[] = [
  { x: 45, y: -15 },
  { x: 150, y: -10 },
  { x: 95, y: -25 },
  { x: 160, y: -30 },
  { x: 40, y: -28 },
];
export const SACCADE_INTERVAL_MS = 1300;
export const SACCADE_EASE = 9;

export const WANDER_TARGETS: readonly GazePoint[] = [
  { x: 70, y: 60 },
  { x: 130, y: 40 },
  { x: 100, y: 72 },
  { x: 55, y: 30 },
  { x: 148, y: 62 },
];
export const WANDER_INTERVAL_MS = 4100;
export const WANDER_EASE = 2.2;
export const ANCHOR_EASE = 6;

export const REST_GAZE: GazePoint = { x: 100, y: 66 };

export function lerp(from: number, to: number, t: number): number {
  return from + (to - from) * t;
}

export function approach(current: number, target: number, rate: number, dt: number): number {
  return current + (target - current) * Math.min(1, rate * dt);
}

export function parseActivity(value: string | null): AgentActivity | null {
  if (value === null) return null;
  return (ACTIVITIES as readonly string[]).includes(value) ? (value as AgentActivity) : 'idle';
}

export function parseDrowseDelay(value: string | null): number {
  const parsed = Number.parseFloat(value ?? '');
  return Number.isFinite(parsed) && parsed >= 0 ? parsed : DEFAULT_DROWSE_DELAY_S;
}

export function shapeTargetFor(activity: AgentActivity | null): number {
  return activity === 'working' ? 1 : 0;
}

export function socketRx(shape: number): number {
  return lerp(EYE_R, SOCKET_MIN_RX, shape);
}

export function pupilRx(radius: number, shape: number): number {
  return lerp(radius, radius * PUPIL_MIN_FRACTION, shape);
}

export function fillToPupilScale(fill: number): number {
  if (fill <= 50) return 1;
  if (fill >= 85) return 2.2;
  return 1 + ((fill - 50) / 35) * 1.2;
}

export function popScale(remainingMs: number): number {
  return remainingMs <= 0 ? 1 : 1 + POP_GAIN * Math.min(1, remainingMs / POP_MS);
}

export function travelClamp(pupilRadius: number): number {
  return Math.max(2, Math.min(MAX_OFFSET, EYE_R - pupilRadius - 4));
}

export function topLidY(fraction: number): number {
  return fraction > LID_OPEN_EPSILON
    ? EYE_CY - EYE_R + fraction * 2 * EYE_R
    : EYE_CY - EYE_R - LID_OVERSHOOT;
}

export function bottomLidY(fraction: number): number {
  return fraction > LID_OPEN_EPSILON
    ? EYE_CY + EYE_R - fraction * 2 * EYE_R
    : EYE_CY + EYE_R + LID_OVERSHOOT;
}

export function chordHalfWidth(y: number, shape: number): number {
  const dy = y - EYE_CY;
  const round = Math.sqrt(Math.max(0, EYE_R * EYE_R - dy * dy));
  return lerp(round, EYE_R - 2, shape);
}

export function drowseLid(awaitingSeconds: number, delaySeconds: number): number {
  if (awaitingSeconds <= delaySeconds) return DROWSE_START_LID;
  const t = Math.min(1, (awaitingSeconds - delaySeconds) / DROWSE_RAMP_S);
  return lerp(DROWSE_START_LID, DROWSE_END_LID, t);
}

export function isLeftRaised(pair: BrowPair): boolean {
  return pair.left.raise < 0;
}

export function recockBrows(previous: BrowPair, random: () => number = Math.random): BrowPair {
  const leftRaised =
    random() < RECOCK_FLIP_CHANCE ? !isLeftRaised(previous) : isLeftRaised(previous);
  const raised: BrowPose = { raise: -(7 + random() * 5), tilt: 7 + random() * 5 };
  const settled: BrowPose = { raise: 1 + random() * 2, tilt: 4 + random() * 3 };

  return leftRaised
    ? { left: { raise: raised.raise, tilt: -raised.tilt }, right: settled }
    : { left: { raise: settled.raise, tilt: -settled.tilt }, right: raised };
}

export function nextGazeIndex(current: number, length: number, random: () => number): number {
  if (length < 2) return 0;
  return (current + 1 + Math.floor(random() * (length - 1))) % length;
}

File: packages/ios-app/SliccFollower/Models/AvatarExpression.swift
import Foundation

enum AvatarExpression {

    enum Activity: String, Equatable, Sendable, Codable, CaseIterable {
        case idle
        case thinking
        case working
        case awaiting
    }

    struct GazePoint: Equatable, Sendable, Codable {
        var x: Double
        var y: Double
    }

    struct BrowPose: Equatable, Sendable, Codable {

        var raise: Double

        var tilt: Double
    }

    struct BrowPair: Equatable, Sendable, Codable {
        var left: BrowPose
        var right: BrowPose
    }

    static let eyeRadius = 38.0
    static let eyeCenterY = 50.0
    static let leftEyeX = 55.0
    static let rightEyeX = 145.0
    static let pupilRadius = 18.0
    static let maxOffset = 16.0

    static let socketMinRx = 10.0

    static let pupilMinFraction = 0.22

    static let shapeEase = 6.0

    static let blinkInSeconds = 0.110
    static let blinkOutSeconds = 0.130

    static let blinkApexSeconds = 0.120
    static let blinkSquish = 0.08
    static let blinkPeriodLeftSeconds = 3.4
    static let blinkPeriodRightSeconds = 4.6

    static let lidEase = 5.0

    static let glowerLid = 0.38
    static let glowerSeconds = 2.6

    static let scrutinyLid = 0.22
    static let scrutinySeconds = 1.0
    static let drowseStartLid = 0.1
    static let drowseEndLid = 0.55
    static let drowseRampSeconds = 12.0
    static let defaultDrowseDelaySeconds = 90.0

    static let lidOpenEpsilon = 0.001

    static let lidLineEpsilon = 0.02

    static let lidOvershoot = 3.0

    static let popSeconds = 0.35
    static let popGain = 0.16

    static let browHalfWidth = 22.0
    static let browY = 2.0
    static let browStroke = 8.0
    static let browTransitionSeconds = 0.35

    static let recockFlipChance = 0.65

    static let baseBrows = BrowPair(
        left: BrowPose(raise: -9, tilt: -10),
        right: BrowPose(raise: 2, tilt: 6))

    static let saccadeTargets: [GazePoint] = [
        GazePoint(x: 45, y: -15),
        GazePoint(x: 150, y: -10),
        GazePoint(x: 95, y: -25),
        GazePoint(x: 160, y: -30),
        GazePoint(x: 40, y: -28),
    ]
    static let saccadeIntervalSeconds = 1.3
    static let saccadeEase = 9.0

    static let wanderTargets: [GazePoint] = [
        GazePoint(x: 70, y: 60),
        GazePoint(x: 130, y: 40),
        GazePoint(x: 100, y: 72),
        GazePoint(x: 55, y: 30),
        GazePoint(x: 148, y: 62),
    ]
    static let wanderIntervalSeconds = 4.1
    static let wanderEase = 2.2
    static let anchorEase = 6.0

    static let restGaze = GazePoint(x: 100, y: 66)

    static func lerp(_ from: Double, _ to: Double, _ t: Double) -> Double {
        from + (to - from) * t
    }

    static func approach(
        current: Double, target: Double, rate: Double, dt: TimeInterval
    ) -> Double {
        current + (target - current) * min(1, rate * dt)
    }

    static func parseActivity(_ raw: String?) -> Activity? {
        guard let raw else { return nil }
        return Activity(rawValue: raw) ?? .idle
    }

    static func shapeTarget(for activity: Activity?) -> Double {
        activity == .working ? 1 : 0
    }

    static func socketRx(shape: Double) -> Double {
        lerp(eyeRadius, socketMinRx, shape)
    }

    static func pupilRx(radius: Double, shape: Double) -> Double {
        lerp(radius, radius * pupilMinFraction, shape)
    }

    static func fillToPupilScale(_ fill: Double?) -> Double {
        guard let fill else { return 1 }
        let clamped = min(100, max(0, fill))
        if clamped <= 50 { return 1 }
        if clamped >= 85 { return 2.2 }
        return 1 + ((clamped - 50) / 35) * 1.2
    }

    static func popScale(remaining: TimeInterval) -> Double {
        remaining <= 0 ? 1 : 1 + popGain * min(1, remaining / popSeconds)
    }

    static func travelClamp(pupilRadius: Double) -> Double {
        max(2, min(maxOffset, eyeRadius - pupilRadius - 4))
    }

    static func topLidY(fraction: Double) -> Double {
        fraction > lidOpenEpsilon
            ? eyeCenterY - eyeRadius + fraction * 2 * eyeRadius
            : eyeCenterY - eyeRadius - lidOvershoot
    }

    static func bottomLidY(fraction: Double) -> Double {
        fraction > lidOpenEpsilon
            ? eyeCenterY + eyeRadius - fraction * 2 * eyeRadius
            : eyeCenterY + eyeRadius + lidOvershoot
    }

    static func chordHalfWidth(y: Double, shape: Double) -> Double {
        let dy = y - eyeCenterY
        let round = (max(0, eyeRadius * eyeRadius - dy * dy)).squareRoot()
        return lerp(round, eyeRadius - 2, shape)
    }

    static func drowseLid(awaiting: TimeInterval, delay: TimeInterval) -> Double {
        guard awaiting > delay else { return drowseStartLid }
        let t = min(1, (awaiting - delay) / drowseRampSeconds)
        return lerp(drowseStartLid, drowseEndLid, t)
    }

    static func isLeftRaised(_ pair: BrowPair) -> Bool {
        pair.left.raise < 0
    }

    static func recockBrows(
        previous: BrowPair,
        random: () -> Double = { Double.random(in: 0..<1) }
    ) -> BrowPair {
        let leftRaised = random() < recockFlipChance ? !isLeftRaised(previous) : isLeftRaised(previous)
        let raised = BrowPose(raise: -(7 + random() * 5), tilt: 7 + random() * 5)
        let settled = BrowPose(raise: 1 + random() * 2, tilt: 4 + random() * 3)

        return leftRaised
            ? BrowPair(left: BrowPose(raise: raised.raise, tilt: -raised.tilt), right: settled)
            : BrowPair(left: BrowPose(raise: settled.raise, tilt: -settled.tilt), right: raised)
    }

    static func nextGazeIndex(current: Int, count: Int, random: () -> Double) -> Int {
        guard count >= 2 else { return 0 }
        return (current + 1 + Int(random() * Double(count - 1))) % count
    }
}
